File: 3014-minimum-number-of-pushes-to-type-word-i/3014-minimum-number-of-pushes-to-type-word-i.c
#include <string.h>
#include <stdlib.h>

// Comparator for quick-sorting frequencies in descending order
int cmpDesc(const void* a, const void* b) {
    return (*(int*)b - *(int*)a);
}

int minimumPushes(char* word) {
    int counts[26] = {0};
    
    // Step 1: Count character frequencies
    for (int i = 0; word[i] != '\0'; i++) {
        counts[word[i] - 'a']++;
    }
    
    // Step 2: Sort the fixed 26-element array in descending order
    qsort(counts, 26, sizeof(int), cmpDesc);
    
    // Step 3: Sum the push requirements based on key availability tiers
    int totalPushes = 0;
    for (int i = 0; i < 26; i++) {
        if (counts[i] == 0) break;
        totalPushes += counts[i] * ((i / 8) + 1);
    }
    
    return totalPushes;
}

File: 3014-minimum-number-of-pushes-to-type-word-i/3014-minimum-number-of-pushes-to-type-word-i.js
/**
 * @param {string} word
 * @return {number}
 */
var minimumPushes = function(word) {
    // Step 1: Count character frequencies
    const counts = new Int32Array(26);
    for (let i = 0; i < word.length; i++) {
        counts[word.charCodeAt(i) - 97]++;
    }
    
    // Step 2: Sort frequencies in descending order
    counts.sort().reverse();
    
    // Step 3: Compute optimal pushes greedily based on 8-key mapping tiers
    let totalPushes = 0;
    for (let i = 0; i < 26; i++) {
        if (counts[i] === 0) break; // Optimization: stop early if no more characters
        totalPushes += counts[i] * (Math.floor(i / 8) + 1);
    }
    
    return totalPushes;
};
